### PR TITLE
AP_HAL_PX4: use new I2CDevice interface

### PR DESCRIPTION
**Pull request by lgywata**
@lucasdemarchi

### CODE REVIEW

**Comment by kwikius** [comment]
Is this interface intended to read and write to eeprom such as I believe is used on OpenPilot Revo?

http://www.microchip.com/wwwproducts/en/24LC128

If so how is it designed to work with it?

**Comment by lucasdemarchi** [comment]
It is intended to replace what we currently have for SPI and I2C, yes. And we add missing features to it as there's a need. What is there in this particular chip that would not work with the way it's implemented?

**Comment by lucasdemarchi** [comment]
Just as  note here that I'm holding the PR until we allow to easily run the drivers in a separate thread.

**Comment by kwikius** [comment]
I recently trialled reading/writing an I2C eeprom (the advantages are very cheap and small size compared to fram and 2 wire i2c bus is also compact). The algorithm to read  requires writing the write address in eeprom to the chip then sending so-called repeated start condition ( ie no stop in between) and starting read cycle (now sending read address then reading data) without giving up the bus. It is somewhat more complex and low level than say reading/writing mag or a baro registers.

 It is no big deal, but If you want a generic device then it would be worth trying to read write the i2c eeprom I cited. I think getting that working that would verify it as a generic i2c driver

**Comment by lucasdemarchi** [comment]
@kwikius if anyone wants to contribute a change to support it, it's ok.

For Linux it depends on the adapter supporting multiple start conditions. I don't think this is so common. Usually multiple start condition is needed when you want to get the slave unstuck or change address (if the part doesn't have a dedicated pin for that). But in the later nothing impedes you to use the default address.

File: libraries/AP_HAL_PX4/I2CDevice.cpp [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+bool I2CDevice::transfer(const uint8_t *send, uint32_t send_len,
+                         uint8_t *recv, uint32_t recv_len)
+{
+    return _device.do_transfer(_address, send, send_len, recv, recv_len);
+}
+
+bool I2CDevice::read_registers_multiple(uint8_t first_reg, uint8_t *recv,
+                                        uint32_t recv_len, uint8_t times)
+{
+    return true;
```
Comment (lucasdemarchi) [inline libraries/AP_HAL_PX4/I2CDevice.cpp]: why is this not implemented?  should at least return false in this case, but implementing it is reasonable.

**Comment by lucasdemarchi** [comment]
@lgywata small comment above, but this looks good.

File: libraries/AP_HAL_PX4/I2CDevice.cpp (hunk 1 above)
Comment (lgywata) [inline libraries/AP_HAL_PX4/I2CDevice.cpp]: Ok, thanks. I'll implement it.

**Comment by lgywata** [comment]
see v2 #4078

**Pull request closed without merging**